### PR TITLE
[Feature](Variant) Implement variant new sub column access method

### PR DESCRIPTION
**Pull request by eldenmoon**
The query SELECT v["a"]["b"] from simple_var WHERE cast(v["a"]["b"] as int) = 1 encompasses three primary testing scenarios:

```
1. A basic test involving the variant data type.
2. A scenario dealing with GitHub event data in the context of a variant.
3. A case related to the TPC-H benchmark using a variant.
```

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by hello-stephen** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        return !result.isEmpty();
+    }
+
+    private static void getElementAtFunction(Expr expr, List<Expr> result) {
+        if (isElementAtOfVariantType(expr)) {
+            result.add(expr);
+            return;
+        }
+        for (Expr child : expr.getChildren()) {
+            if (isElementAtOfVariantType(expr)) {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java:93]: The following recusive call is enough.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        return slot;
+    }
+
+    private static boolean isElementAtOfVariantType(Expr expr) {
+        if (!(expr instanceof FunctionCallExpr)) {
+            return false;
+        }
+        FunctionCallExpr functionCallExpr = (FunctionCallExpr) expr;
+        return functionCallExpr.getFnName().getFunction().equalsIgnoreCase("element_at")
+                && functionCallExpr.getType() == Type.VARIANT;
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java:78]: functionCallExpr.getType() is return type of function. I think the type of the first arg should be checked instead.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    private Expr apply(Expr expr, Analyzer analyzer) throws AnalysisException  {
+        if (!(isElementAtOfVariantType(expr))) {
+            return expr;
+        }
+        List<SlotRef> slotRefs =  Lists.newArrayList();
+        expr.collect(SlotRef.class, slotRefs);
+        SlotRef slot = slotRefs.get(0);
+        List<Expr> pathsExpr = Lists.newArrayList();
+        expr.collect(Expr::isLiteral, pathsExpr);
+        List<String> fullPaths = pathsExpr.stream()
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java:64]: Do you mean this transformation: v['a']['b']['c'] => element_at(v, 'a', 'b', 'c') => SlotRef(v, ['a', 'b', 'c']) ?

I assume that v['a']['b']['c'] is analyzed to element_at(element_at(element_at(v, 'a'), 'b'), 'c') originally.

Comment (xiaokang) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java:64]: Add an example comment to explain here.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        List<Expr> originalFunctionElementAtExprs = Lists.newArrayList();
+        Expr newExpr = null;
+        getElementAtFunction(inputExpr, originalFunctionElementAtExprs);
+        for (Expr expr : originalFunctionElementAtExprs) {
+            Expr rewriteExpr = apply(expr, analyzer);
+            if (rewriteExpr != expr) {
+                if (newExpr == null) {
+                    newExpr = inputExpr.clone();
+                }
+                newExpr = replaceExpr(newExpr, expr.getId().toString(), rewriteExpr);
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java:113]: Why not just use expr, but create newExpr?

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**